Add container breakouts security advisories (#3)
Mirantis/security: advisories/0010.md

`Note` Please be aware that this is an `EXCEPTIONAL` and `TEMPORARY` solution and no new ones are expected in the future.
#### Ubuntu
```
# apt-get install containerd.io=1.6.28~rc.1-2*
```
#### RPM package based OSes
```
# yum install containerd.io-1.6.28-4.1.rc.1.1*
```

### Offline deployment
1. Download packages from [official Mirantis repository][mcr.repos] according your OS and platform
2. Copy them to offline nodes (using USB Flash for example)
3. Install these packages on hosts and check `runc` version
#### Ubuntu
```
# dpkg -i docker-ee_23.0.9~4-0* docker-ee-cli_23.0.9~4-0* docker-ee-rootless-extras_23.0.9~4-0* containerd.io=1.6.28~rc.1-2*
```
#### RPM package based OSes
```
# rpm -i docker-ee-23.0.9-4* docker-ee-cli-23.0.9-4* docker-ee-rootless-extras-23.0.9-4* containerd.io-1.6.28-4.1.rc.1.1*
```

### Upgrade MCR using Launchpad 
If your MKE cluster is deployed with Launchpad and configuration correspond to current cluster configuration you can 
update MCR w/ Launchpad using the follow steps:

> 1. Download the [Launchpad][launchpad] v1.5.4 version of later

Reviewer: Small typo :) , I think it should be `or later` instead of `of later`

Author: Done

2. Run Launchpad apply with `--force upgrade` flag:
   - `./launchpad-linux-<arch> apply --force-upgrade` 

> 3. Check `runc` version on all nodes

Reviewer 2: These launchpad steps are ONLY needed if the customer is already using 23.0.9 and needs to upgrade to 23.0.9+security fix.

I think that your text looks good

## MCR: runc process.cwd and leaked fds container breakout 
### Release Date
2024-01-31
### Overview
Several container breakouts due to internally leaked fds.

Consult the [upstream advisory][upstream] and [snyk][snyk.io] for details, mitigations, workarounds, and patches.
### Affected Products
Mirantis Container Runtime (MCR) <= 23.0.9
## Vulnerability Information
### CVE Identifier
[CVE-2024-21626][nvd.nist.gov]
### CVSSv3.1
8.6 (High) CVSS:3.1/AV:L/AC:L/PR:N/UI:R/S:C/C:H/I:H/A:H
### CWEs
CWE-403, CWE-668
### How to check if your environment is vulnerable
`runc` version is equal or lower (<=) than `1.1.11-rc1`
````
# docker version --format '{{range .Server.Components}}{{if eq .Name "runc"}}{{.Version}}{{end}}{{end}}'
1.1.11-rc1 <<<<<<<<<<<<<<<
````
## How to fix
### MCR 23.0.x
Upgrade MCR packages to the fixed 23.0.9-1 version or newer
* `docker-ee*` packages to:
  - `23.0.9~4-0` Ubuntu
  - `23.0.9-4` for RPM package based OSes
* `containerd.io` package to
  - `1.6.28~rc.1-2` Ubuntu
  - `1.6.28-4.1.rc.1.1` for RPM package based OSes
* `Note:` Restart of the `dockerd` service is `NOT` necessary! 
#### Ubuntu
```

> # apt-get install docker-ee=5:23.0.9~4-0* docker-ee-cli=5:23.0.9~4-0* docker-ee-rootless-extras=5:23.0.9~4-0* containerd.io=1.6.28~rc.1-2*

Reviewer 3: ```suggestion
# apt-get install 'docker-ee=5:23.0.9~4-0*' 'docker-ee-cli=5:23.0.9~4-0*' 'docker-ee-rootless-extras=5:23.0.9~4-0*' 'containerd.io=1.6.28~rc.1-2*'
```
We need to tell them something so the asterisks don't get shell-expanded. If you don't like the single quotes, backslash-escape them.

```
Check `runc` version:
```
# docker version --format '{{range .Server.Components}}{{if eq .Name "runc"}}{{.Version}}{{end}}{{end}}'
1.1.11-rc1.m1 <<<<<<<<<<<<<<<
```
#### RPM package based OSes
```

> # yum install docker-ee-3:23.0.9-4* docker-ee-cli-1:23.0.9-4* docker-ee-rootless-extras-23.0.9-4* containerd.io-1.6.28-4.1.rc.1.1*

Reviewer 3: ```suggestion
# yum install 'docker-ee-3:23.0.9-4*' 'docker-ee-cli-1:23.0.9-4*' 'docker-ee-rootless-extras-23.0.9-4*' 'containerd.io-1.6.28-4.1.rc.1.1*'
```

```
Check `runc` version:
```
# docker version --format '{{range .Server.Components}}{{if eq .Name "runc"}}{{.Version}}{{end}}{{end}}'
1.1.11-rc1.m1 <<<<<<<<<<<<<<<
```

For more details please read the [MCR official installation documentation][mcr.docs]

### MCR 20.10.x 
If you have EOL MCR 20.10.x we `STRONGLY RECOMMEND` to upgrade to the `MCR 23.0.9-1`.

> As a temporary solution for environments running `MCR 20.10.12` or newer, the CVE can be resolved by upgrading only the `containerd.io` package to the one shipped with `MCR 23.0.9-1` using the instructions below. This workaround is only possible on `MCR 20.10.12` or newer because this is the first version of MCR 20.10 which consumed containerd 1.6."

Reviewer 4: You have a trailing quote here

Author: Done

`Note` Please be aware that this is an `EXCEPTIONAL` and `TEMPORARY` solution and no further updates are expected.
#### Ubuntu
```

> # apt-get install containerd.io=1.6.28~rc.1-2*

Reviewer 3: ```suggestion
# apt-get install 'containerd.io=1.6.28~rc.1-2*'
```

```
#### RPM package based OSes
```

> # yum install containerd.io-1.6.28-4.1.rc.1.1*

Reviewer 3: ```suggestion
# yum install 'containerd.io-1.6.28-4.1.rc.1.1*'
```

```

### Offline deployment
1. Download packages from [official Mirantis repository][mcr.repos] according your OS and platform
2. Copy them to offline nodes (using USB Flash for example)
3. Install these packages on hosts and check `runc` version
#### Ubuntu
```
# dpkg -i docker-ee_23.0.9~4-0* docker-ee-cli_23.0.9~4-0* docker-ee-rootless-extras_23.0.9~4-0* containerd.io=1.6.28~rc.1-2*
```
#### RPM package based OSes
```
# rpm -i docker-ee-23.0.9-4* docker-ee-cli-23.0.9-4* docker-ee-rootless-extras-23.0.9-4* containerd.io-1.6.28-4.1.rc.1.1*

> ```

Reviewer 3: Shell-expansion is _intended_ here. DO NOT ADD QUOTES OR BACKSLASH ESCAPES

### Upgrade MCR using Launchpad 
If your MKE cluster is deployed with Launchpad and configuration correspond to current cluster configuration you can 
update MCR w/ Launchpad using the follow steps:
1. Download the [Launchpad][launchpad] v1.5.4 version or newer

> 2. Run Launchpad apply with `--force upgrade` flag:

Reviewer 3: ```suggestion
2. Run Launchpad apply with `--force-upgrade` flag:
```

3. Install these packages on hosts and check `runc` version
#### Ubuntu
```
# dpkg -i docker-ee_23.0.9~4-0*.deb docker-ee-cli_23.0.9~4-0*.deb docker-ee-rootless-extras_23.0.9~4-0*.deb containerd.io_1.6.28~rc.1-2*.deb
```
#### RPM package based OSes
```
# rpm -i docker-ee-23.0.9-4*.rpm docker-ee-cli-23.0.9-4*.rpm docker-ee-rootless-extras-23.0.9-4*.rpm containerd.io-1.6.28-4.1.rc.1.1*.rpm
```

### Upgrade MCR using Launchpad 
If your MKE cluster is deployed with Launchpad and configuration correspond to current cluster configuration you can 
update MCR w/ Launchpad using the follow steps:
1. Download the [Launchpad][launchpad] v1.5.4 version or newer
2. Configure MCR version `23.0.9`
For example:
```
apiVersion: launchpad.mirantis.com/mke/v1.3
kind: mke
metadata:
....
spec:
....
  "mcr":
    "channel": "stable"
    "installURLLinux": "https://get.mirantis.com/"
    "installURLWindows": "https://get.mirantis.com/install.ps1"
    "repoURL": "https://repos.mirantis.com"
    "version": "23.0.9"       <<<<<<<<<<<<<<<
....
```
`Note` MCR version must be exactly `23.0.9` (NOT 23.0.9-1)
3. Run Launchpad apply with `--force-upgrade` flag:

>    - `./launchpad-linux-<arch> apply --force-upgrade` 

Reviewer 5: If you look at our public facing docs (https://docs.mirantis.com/mke/3.7/launchpad/lp-get-started.html#install-launchpad) we tell folks to rename the binary to launchpad.  I'd recommend we stay consistent with those docs here in the advisory.

Author: Done

### MCR: runc process.cwd and leaked fds container breakout 
#### Release Date
2024-01-31
#### Overview
Several container breakouts due to internally leaked fds.

Consult the [upstream advisory][upstream] and [snyk][snyk.io] for details, mitigations, workarounds, and patches.
#### Affected Products
Mirantis Container Runtime (MCR) <= 23.0.9
### Vulnerability Information
#### CVE Identifier
[CVE-2024-21626][nvd.nist.gov]
#### CVSSv3.1
8.6 (High) CVSS:3.1/AV:L/AC:L/PR:N/UI:R/S:C/C:H/I:H/A:H
#### CWEs
CWE-403, CWE-668

> #### How to check if your environment is vulnerable

Reviewer 3: They could also check using `docker version`. We should tell them how.

Reviewer 3: ```
docker version --format '{{range .Server.Components}}{{if eq .Name "runc"}}{{.Version}}{{end}}{{end}}'
```

Author: Done

* `containerd.io` package version is equal or lower (<=) than:
  - 1.6.28~rc.1-1 for DEB package based OSes
  - 1.6.28-2.1.rc.1.1 for RPM package based OSes
#### How to fix
Upgrade MCR packages to the fixed 23.0.9 version or newer
* `docker-ee*` packages to:
  - 23.0.9~4-0 for DEB package based OSes

>   - 23.0.9-4 for RPM package based OSes

Reviewer 3: Is an RPM package update really required? [Our RPM spec's requirements](https://github.com/Mirantis/ee-packaging/blob/7dab2c633206f2c6fdd27fe7864986c454a45b68/rpm/SPECS/docker-ee.spec#L50) only [constrain the dependency `Version`, not the `Release`](http://ftp.rpm.org/api/4.4.2.2/dependencies.html), in contrast to our deb packages. cc @ameyag

Reviewer 6: In the absence of `Release`, it'll end up clobbering the last 23.0.9 patch.

Reviewer 3: I'm asking if _users_ need to install the bumped docker-ee rpm

Author: no, but I would prefer to do this to have more consistent state

* `containerd.io` package to
  - 1.6.28~rc.1-2 for DEB package based OSes
  - 1.6.28-4.1.rc.1.1 for RPM package based OSes

> Please follow the [MCR official installation documentation][mcr.docs]

Reviewer 3: Should we mention that a daemon restart is not required?

Author: Done

#### Disclosure Timeline
* 2023-11-20: Rory McNamara initially discovered the vulnerabilities. He began the internal verification process and additional research to validate findings and build POC exploits.
* 2023-12-11: Initial disclosure sent to Docker with all vulnerabilities, and Docker ACKed the same day.
* 2023-12-12: Snyk received a request from Docker to forward the WORKDIR vulnerability to runc, as it was deemed their responsibility.
* 2023-12-13: Rory was added as a Github Security Advisory (GHSA) collaborator for the Arbitrary Delete and grpc Docker/Buildkit vulns (both initially opened 11-Dec-2023).
* 2023-12-19: Rory was added as GHSA collaborator to WORKDIR by runc (initially opened 11-Dec-2023).
* 2023-12-20: Rory was added as GHSA collaborator for the cache race vulnerability.
* 2024-01-02: runc CVE assigned (Github CNA).
* 2024-01-17: runc sends an announcement to their security mailing list including the patches & embargo date of 31-Jan-2024.
* 2024-01-24: Docker vulnerabilities CVEs assigned (GitHub CNA).
* 2024-01-31: All four "Leaky Vessels" vulnerabilities announced publicly.
* 2024-01-31: Runc released version 1.1.12 which fixes the vulnerabilities.

> * 2024-02-08: MCR 23.0.9-1 released

Reviewer 3: ```suggestion
* 2024-02-08: MCR 23.0.9-4 released
```

## MCR: runc process.cwd and leaked fds container breakout 
### Release Date
2024-01-31
### Overview
Several container breakouts due to internally leaked fds.

Consult the [upstream advisory][upstream] and [snyk][snyk.io] for details, mitigations, workarounds, and patches.
### Affected Products
Mirantis Container Runtime (MCR) <= 23.0.9
## Vulnerability Information
### CVE Identifier
[CVE-2024-21626][nvd.nist.gov]
### CVSSv3.1
8.6 (High) CVSS:3.1/AV:L/AC:L/PR:N/UI:R/S:C/C:H/I:H/A:H
### CWEs
CWE-403, CWE-668
### How to check if your environment is vulnerable
`runc` version is equal or lower (<=) than `1.1.11-rc1`
````
# docker version --format '{{range .Server.Components}}{{if eq .Name "runc"}}{{.Version}}{{end}}{{end}}'
1.1.11-rc1 <<<<<<<<<<<<<<<
````

> ## How to fix

Reviewer 3: If we're instructing users to update using their distro package manager, why are we going through all the trouble to fix up launchpad? And I kept hearing in the patch calls that the decision was made to instruct users to go through the normal upgrade process. Did that change?

Upgrade MCR packages to the fixed 23.0.9-1 version or newer
* `docker-ee*` packages to:
  - `23.0.9~4-0` for DEB package based OSes
  - `23.0.9-4` for RPM package based OSes
* `containerd.io` package to
  - `1.6.28~rc.1-2` for DEB package based OSes
  - `1.6.28-4.1.rc.1.1` for RPM package based OSes

> * `Caution:` Restart of the `dockerd` service is `NOT` needed! 

Reviewer 3: ```suggestion
* `Note:` Restart of the `dockerd` service is `NOT` necessary! 
```

Author: Done

> #### DEB package based OSes

Reviewer 3: Wait a sec... we only build DEBs for Ubuntu, not Debian or any other Debian-derived distro.
```suggestion
#### Ubuntu
```

Author: Done

```

> # sudo apt-get install docker-ee=5:23.0.9~4-0~<OS> docker-ee-cli=5:23.0.9~4-0~<OS> docker-ee-rootless-extras=5:23.0.9~4-0~<OS> containerd.io=1.6.28~rc.1-2*

Reviewer 3: `#` implies root shell, making `sudo` redundant. Why is a wildcard used to elide the OS version for the containerd package but not for any of the others? And asterisks in the shell without escaping is a dangerous game to play.
```suggestion
# apt-get install 'docker-ee=5:23.0.9~4-0*' 'docker-ee-cli=5:23.0.9~4-0*' 'docker-ee-rootless-extras=5:23.0.9~4-0*' 'containerd.io=1.6.28~rc.1-2*'
```

```console
root@049138adc5fb:/# apt install --dry-run docker-ee=5:23.0.9~3.4rc1\*
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Selected version '5:23.0.9~3.4rc1-0~ubuntu-jammy' (Mirantis Container Runtime:jammy [amd64]) for 'docker-ee'
```

Author: Done

```
Check `runc` version:
```
# docker version --format '{{range .Server.Components}}{{if eq .Name "runc"}}{{.Version}}{{end}}{{end}}'
1.1.11-rc1.m1 <<<<<<<<<<<<<<<
```

> #### RPM package based OSes

Reviewer 3: Not all RPM based OSes use yum. SLES/OpenSUSE uses `zypper` and, judging by the special case in get.mirantis.com/install.sh, probably can't use yum.

Author: you are right, but haven't checked that

```

> # sudo yum install docker-ee-3:23.0.9-4.<OS>.<Arch> docker-ee-cli-1:23.0.9-4.<OS>.<Arch> docker-ee-rootless-extras-23.0.9-4.<OS>.<Arch> containerd.io-1.6.28-4.1.rc.1.1.<OS>.<Arch>

Reviewer 3: Wildcards work here, too.
```suggestion
# yum install 'docker-ee-3:23.0.9-4.*' 'docker-ee-cli-1:23.0.9-4.*' 'docker-ee-rootless-extras-23.0.9-4.*' 'containerd.io-1.6.28-4.1.rc.1.1.*'
```

Author: Done

```
Check `runc` version:
```
# docker version --format '{{range .Server.Components}}{{if eq .Name "runc"}}{{.Version}}{{end}}{{end}}'
1.1.11-rc1.m1 <<<<<<<<<<<<<<<
```

For more details please read the [MCR official installation documentation][mcr.docs]

### MCR 20.10.x 
If you have MCR 20.10.12 (`containerd` 1.6 is used starting from 20.10.12) please upgrade `containerd.io` package to the fixed version:
#### DEB package based OSes
```
# sudo apt-get install containerd.io=1.6.28~rc.1-2*
```
#### RPM package based OSes
```
# sudo yum install containerd.io-1.6.28-4.1.rc.1.1.<OS>.<Arch>

> ```

Reviewer 3: IIRC the 20.10 instructions are supposed to be passed along to Mirantis CSO but not published to docs.mirantis.com since 20.10 is EOL and we don't want to set precedent

Author: Correct, but then it was decided to publish as well
I've updated the message a bit, please review

### Offline deployment
1. Download packages from [official Mirantis repository][mcr.repos] according your OS and platform
2. Copy them to offline nodes (using USB Flash for example)

> 3. Install these packages on hosts and check `crun` version

Reviewer 3: ```suggestion
3. Install these packages on hosts and check `runc` version
```

Author: Done

For more details please read the [MCR official installation documentation][mcr.docs]

### MCR 20.10.x 
If you have EOL MCR 20.10.x we `STRONGLY RECOMMEND` to upgrade to the MCR 23.0.9-1.

> But we have an temporary option in case if the MCR version is 20.10.12 (`containerd` 1.6 is used starting from 20.10.12) or newer the CVE can be resolved by upgrade `containerd.io` package to the fixed version.

Reviewer 4: Suggest change to

As a temporary solution for environments running MCR 20.10.12 or newer, the CVE can be resolved by upgrading only the `containerd.io` package to the one shipped with MCR 23.0.9-1 using the instructions below. This workaround is only possible on MCR 20.10.12 or newer because this is the first version of MCR 20.10 which consumed `containerd` 1.6."

Author: Done

> `Note` Please be aware that this is an `EXCEPTIONAL` and `TEMPORARY` solution and no new ones are expected in the future.

Reviewer 4: "and no new ones are expected" -> "and no further updates are expected" ?

Author: Done
